nengo/builder/optimizer.py

```diff
@@ -11,386 +10,707 @@
+        # performed on at most one type of operator per pass.
+        # The depedency graph and other information will be updated
+        # before merging other operator types.
+
+        # We go through ops in a heuristic order to reduce runtime
+        firstops = [ElementwiseInc, SlicedCopy, DotInc, SimNeurons]
+        sortedops = firstops + [op for op in by_type if op not in firstops]
+        for optype in sortedops:
+
+            if _OpMerger.is_type_mergeable(optype):
+                self.perform_merges_for_subset(by_type[optype])
+
+            # If we're not only merging views, the memory layout changes
+            # and non-views are turned into views. In that case we need
+            # to update the signals the operators are referring to before
+            # trying to merge a different type of operators. Thus, we break
+            # the loop here.
+            if not self.only_merge_ops_with_view and len(self.merged) > 0:
+                break
 
-    @staticmethod
-    def _first_view_base(op):
-        """Returns the base of the first signal view of operator `op`.
-
-        Returns ``None`` if the operator has no signal views."""
-        for s in op.all_signals:
-            if s.is_view:
-                return s.base
-        return None
-
-    def _check_sequential(self, op1, op2):
-        """Checks that the first signal views of `op1` and `op2` are
-        non-overlapping and the view of `op2` starts after the view of `op1`.
-
-        This is similar to the `_is_memory_access_sequential`, but has some
-        minute differences. In contrast to `_is_memory_access_sequential`, this
-        function
-
-        * only checks the first view instead of all views
-        * does not check that all signals are views
-        * only checks two operators at a time instead of a whole list
-        * only returns ``True`` if there is a gap between the end of the view
-          of `op1` and the start of the view of `op2`
-
-        This function is only intended to give a quick indication whether the
-        merge of `op1` and `op2` could potentially be possible to allow for an
-        early break out of the inner loop given a list of sorted operators by
-        the offset of their first signal view; whereas
-        `_is_memory_access_sequential` is used for the final check that all
-        the views match up in their memory alignment.
-        """
-        return (
-            self._first_view_offset(op1) + self._first_view_size(op1) <
-            self._first_view_offset(op2))
+        # At this point, we may have marked some ops for merge.
+        # Some signals may therefore have been replaced.
+        # Those signals might be used in some other op, so we go through
+        # all other ops and do the same signal replacement if necessary.
 
-    def _perform_merges_for_subset(
-            self, subset, tc, only_merge_ops_with_view=True):
+        # TODO: do we have to do all ops or can we get away with a subset?
```

Comment:
Potentially with `sig2ops`?

To me the following piece is a different level of abstraction and should go in its own function.